ARDUINO/LaCOOLBoard/extras/doxygenDocs/man/man3/CoolMQTT.3
.TH "CoolMQTT" 3 "Tue Sep 19 2017" "CoolBoardAPI" \" -*- nroff -*-
.ad l
.nh
.SH NAME
CoolMQTT \- This class handles the mqtt client\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <CoolMQTT\&.h>\fP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "void \fBbegin\fP ()"
.br
.ti -1c
.RI "int \fBconnect\fP (unsigned long keepAlive)"
.br
.ti -1c
.RI "bool \fBpublish\fP (const char *data)"
.br
.ti -1c
.RI "bool \fBpublish\fP (const char *data, unsigned long logInterval)"
.br
.ti -1c
.RI "String \fBread\fP ()"
.br
.ti -1c
.RI "void \fBconfig\fP (const char \fBmqttServer\fP[], const char \fBinTopic\fP[], const char \fBoutTopic\fP[], const char \fBuser\fP[], int \fBbufferSize\fP)"
.br
.ti -1c
.RI "bool \fBconfig\fP ()"
.br
.ti -1c
.RI "void \fBcallback\fP (char *topic, byte *payload, unsigned int length)"
.br
.ti -1c
.RI "void \fBprintConf\fP ()"
.br
.ti -1c
.RI "int \fBstate\fP ()"
.br
.ti -1c
.RI "bool \fBmqttLoop\fP ()"
.br
.ti -1c
.RI "String \fBgetUser\fP ()"
.br
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "char \fBmqttServer\fP [50] ={'0'}"
.br
.ti -1c
.RI "String \fBmsg\fP =''"
.br
.ti -1c
.RI "char \fBinTopic\fP [50] ={'0'}"
.br
.ti -1c
.RI "char \fBoutTopic\fP [50] ={'0'}"
.br
.ti -1c
.RI "char \fBuser\fP [50] ={'0'}"
.br
.ti -1c
.RI "int \fBbufferSize\fP =3000"
.br
.ti -1c
.RI "WiFiClient \fBespClient\fP"
.br
.ti -1c
.RI "\fBCoolPubSubClient\fP \fBclient\fP"
.br
.ti -1c
.RI "bool \fBnewMsg\fP =0"
.br
.ti -1c
.RI "unsigned long \fBpreviousLogTime\fP =0"
.br
.in -1c
.SH "Detailed Description"
.PP 
This class handles the mqtt client\&. 
.PP
Definition at line 45 of file CoolMQTT\&.h\&.
.SH "Member Function Documentation"
.PP 
.SS "void CoolMQTT::begin ()"
\fBCoolMQTT::begin()\fP: This method is provided to set the mqtt client's parameters: -client -server -callback method -buffer size 
.PP
Definition at line 54 of file CoolMQTT\&.cpp\&.
.SS "void CoolMQTT::callback (char * topic, byte * payload, unsigned int length)"
CoolMQTT::callback(in topic, incoming message , message length): This method is provided to handle incoming messages from the subscribed inTopic\&.
.PP
Arguments are automatically assigned in client\&.setCallback() 
.PP
Definition at line 310 of file CoolMQTT\&.cpp\&.
.SS "void CoolMQTT::config (const char mqttServer[], const char inTopic[], const char outTopic[], const char user[], int bufferSize)"
CoolMQTT::config(server,in topic, out topic , user Id, buffer size): This method is provided to manually configure the mqtt client 
.PP
Definition at line 627 of file CoolMQTT\&.cpp\&.
.SS "bool CoolMQTT::config ()"
\fBCoolMQTT::config()\fP: This method is provided to configure the mqttClient : -server -inTopic -outTopic -client Id -buffer size
.PP
\fBReturns:\fP
.RS 4
true if successful,false otherwise 
.RE
.PP

.PP
Definition at line 430 of file CoolMQTT\&.cpp\&.
.SS "int CoolMQTT::connect (unsigned long keepAlive)"
CoolMQTT::connect( time to keep the connection alive in seconds ): This method is provided to connect the client to the server, publish to the out topic , subscribe to the in topic and set the keepAlive time\&.
.PP
\fBReturns:\fP
.RS 4
mqtt client state 
.RE
.PP

.PP
Definition at line 110 of file CoolMQTT\&.cpp\&.
.SS "String CoolMQTT::getUser ()"
\fBCoolMQTT::getUser()\fP: This method is provided to get the user name 
.PP
Definition at line 690 of file CoolMQTT\&.cpp\&.
.SS "bool CoolMQTT::mqttLoop ()"
\fBCoolMQTT::mqttLoop()\fP: This method is provided to allow the client to process the data
.PP
\fBReturns:\fP
.RS 4
true if successful,false otherwise 
.RE
.PP

.PP
Definition at line 274 of file CoolMQTT\&.cpp\&.
.SS "void CoolMQTT::printConf ()"
\fBCoolMQTT::printConf()\fP: This method is provided to print the configuration to the Serial Monitor 
.PP
Definition at line 654 of file CoolMQTT\&.cpp\&.
.SS "bool CoolMQTT::publish (const char * data)"
CoolMQTT::publish(data): This method is provided to publish data to the out topic
.PP
\fBReturns:\fP
.RS 4
true if publish successful, false otherwise 
.RE
.PP

.PP
Definition at line 171 of file CoolMQTT\&.cpp\&.
.SS "bool CoolMQTT::publish (const char * data, unsigned long logInterval)"
CoolMQTT::publish(data): This method is provided to publish data to the out topic every logInterval in seconds
.PP
\fBReturns:\fP
.RS 4
true if publish successful, false otherwise 
.RE
.PP

.PP
Definition at line 222 of file CoolMQTT\&.cpp\&.
.SS "String CoolMQTT::read ()"
\fBCoolMQTT::read()\fP: This method is provided to return the last read message\&. 
.PP
Definition at line 389 of file CoolMQTT\&.cpp\&.
.SS "int CoolMQTT::state ()"
\fBCoolMQTT::state()\fP: This method is provided to return the mqtt client's state\&. 
.PP
\fBReturns:\fP
.RS 4
mqtt client state: -4 : MQTT_CONNECTION_TIMEOUT - the server didn't respond within the keepalive time -3 : MQTT_CONNECTION_LOST - the network connection was broken -2 : MQTT_CONNECT_FAILED - the network connection failed -1 : MQTT_DISCONNECTED - the client is disconnected cleanly 0 : MQTT_CONNECTED - the cient is connected 1 : MQTT_CONNECT_BAD_PROTOCOL - the server doesn't support the requested version of MQTT 2 : MQTT_CONNECT_BAD_CLIENT_ID - the server rejected the client identifier 3 : MQTT_CONNECT_UNAVAILABLE - the server was unable to accept the connection 4 : MQTT_CONNECT_BAD_CREDENTIALS - the username/password were rejected 5 : MQTT_CONNECT_UNAUTHORIZED - the client was not authorized to connect 
.RE
.PP

.PP
Definition at line 87 of file CoolMQTT\&.cpp\&.
.SH "Member Data Documentation"
.PP 
.SS "int CoolMQTT::bufferSize =3000\fC [private]\fP"
MQTT inner bufferSize 
.PP
Definition at line 104 of file CoolMQTT\&.h\&.
.SS "\fBCoolPubSubClient\fP CoolMQTT::client\fC [private]\fP"
MQTT Client instance 
.PP
Definition at line 114 of file CoolMQTT\&.h\&.
.SS "WiFiClient CoolMQTT::espClient\fC [private]\fP"
WifiClient instance 
.PP
Definition at line 109 of file CoolMQTT\&.h\&.
.SS "char CoolMQTT::inTopic[50] ={'0'}\fC [private]\fP"
MQTT Topic to subscribe/listen to 
.PP
Definition at line 89 of file CoolMQTT\&.h\&.
.SS "char CoolMQTT::mqttServer[50] ={'0'}\fC [private]\fP"
MQTT Server name/ip 
.PP
Definition at line 79 of file CoolMQTT\&.h\&.
.SS "String CoolMQTT::msg =''\fC [private]\fP"
String to store incoming messages 
.PP
Definition at line 84 of file CoolMQTT\&.h\&.
.SS "bool CoolMQTT::newMsg =0\fC [private]\fP"
new message flag 
.PP
Definition at line 119 of file CoolMQTT\&.h\&.
.SS "char CoolMQTT::outTopic[50] ={'0'}\fC [private]\fP"
MQTT topic to publish/write to 
.PP
Definition at line 94 of file CoolMQTT\&.h\&.
.SS "unsigned long CoolMQTT::previousLogTime =0\fC [private]\fP"
last time the Client sent a Message over MQTT in ms 
.PP
Definition at line 125 of file CoolMQTT\&.h\&.
.SS "char CoolMQTT::user[50] ={'0'}\fC [private]\fP"
MQTT user name 
.PP
Definition at line 99 of file CoolMQTT\&.h\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for CoolBoardAPI from the source code\&.


ARDUINO/LaCOOLBoard/src/CoolMQTT.cpp
/**
*	Copyright (c) 2018 La Cool Co SAS
*
*	Permission is hereby granted, free of charge, to any person obtaining a copy
*	of this software and associated documentation files (the "Software"), to deal
*	in the Software without restriction, including without limitation the rights
*	to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
*	copies of the Software, and to permit persons to whom the Software is
*	furnished to do so, subject to the following conditions:
*
*	The above copyright notice and this permission notice shall be included in all
*	copies or substantial portions of the Software.
*
*	THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
*	IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
*	FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
*	AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
*	LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
*	OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
*	SOFTWARE.
*
*/



#include "FS.h"
#include "Arduino.h"  
#include <ESP8266WiFi.h>
#include "CoolMQTT.h"
#include "ArduinoJson.h"
#include "CoolWifi.h"


#define DEBUG 0



/**
*	CoolMQTT::begin():
*	This method is provided to set the mqtt
*	client's parameters:	-client
*				-server
*				-callback method
*				-buffer size
*/
void CoolMQTT::begin()
{ 

#if DEBUG == 1 

	Serial.println( F("Entering CoolMQTT.begin()") );
	Serial.println();

#endif

	client.setClient(espClient);
	client.setServer(mqttServer, 1883);	
	client.setCallback([this] (char* topic, byte* payload, unsigned int length) { this->callback(topic, payload, length); });
	client.setBufferSize((unsigned short)bufferSize);

}

/**
*	CoolMQTT::state():
*	This method is provided to return the 
*	mqtt client's state.
*	\return mqtt client state:	
*		-4 : MQTT_CONNECTION_TIMEOUT - the server didn't respond within the keepalive time
*		-3 : MQTT_CONNECTION_LOST - the network connection was broken
*		-2 : MQTT_CONNECT_FAILED - the network connection failed
*		-1 : MQTT_DISCONNECTED - the client is disconnected cleanly
*		0 : MQTT_CONNECTED - the cient is connected
*		1 : MQTT_CONNECT_BAD_PROTOCOL - the server doesn't support the requested version of MQTT
*		2 : MQTT_CONNECT_BAD_CLIENT_ID - the server rejected the client identifier
*		3 : MQTT_CONNECT_UNAVAILABLE - the server was unable to accept the connection
*		4 : MQTT_CONNECT_BAD_CREDENTIALS - the username/password were rejected
*		5 : MQTT_CONNECT_UNAUTHORIZED - the client was not authorized to connect
*/
int CoolMQTT::state()
{

#if DEBUG == 1 

	Serial.println( F("Entering CoolMQTT.state()") );
	Serial.println();	
	Serial.print( F("state : ") );
	Serial.println( this->client.state() );

#endif
	
	return( this->client.state() );
}

/**
*	CoolMQTT::connect( time to keep the connection alive in seconds ):
*	This method is provided to connect the client to the server,
*	publish to the out topic , subscribe to the in topic and set
*	the keepAlive time.
*	
*	\return mqtt client state
*/
int CoolMQTT::connect(unsigned long keepAlive)
{       

	int i=0;

#if DEBUG == 1 

	Serial.println( F("Entering CoolMQTT.connect()") );

#endif
	Serial.println( F("MQTT connecting...") );
	
	while( ( !this->client.connected() ) && ( i<5 ) ) 
	{
		// Attempt to connect
		if( this->client.connect( this-> user,keepAlive*10  ) )
		{
			client.subscribe( this->inTopic );

		#if DEBUG == 1 

			Serial.println( F(" subscribed , leavin ") ) ;
		
		#endif

			return( this->state() );
		}

		else
		{
		
		#if DEBUG == 1 

			Serial.println( F("not connected , retrying") );
		
		#endif

			
		}

	delay(5);
	i++;
	}
	if (state() == 0)
	{
		Serial.println( F("MQTT connected : OK") );
		Serial.println();
	}
	else Serial.println( F("MQTT not jet connected..."));
	return( this->state() );

}

/**
*	CoolMQTT::publish(data):
*	This method is provided to publish data
*	to the out topic. If the publish fails 
*   it deconnects and reconnects the Wifi 
*   and MQTT. This prevents lost MQTT syndrome
*   on cheaper Wifi hotspots
*
*	\return true if publish successful,
*	false otherwise
*/
bool CoolMQTT::publish(const char* data)
{

#if DEBUG == 1 

	Serial.println( F("Entering CoolMQTT.publish()") );
	Serial.println();
	//data is in JSON, publish it directly

	Serial.println( F("data to publish : ") );
	Serial.println(data);
	Serial.print( F("data size : ") );
	Serial.println(strlen(data));

	Serial.println();

#endif
#if DEBUG == 0
	Serial.println( F("Publishing Message : "));
	Serial.println(data);
	Serial.println();
#endif

	bool pub=client.publish( this->outTopic,(byte*) data,strlen(data),false  );

#if DEBUG == 1 

	Serial.print( F("success : ") );
	Serial.println(pub);	

#endif
	if (pub == 1)
	{
		Serial.println( F("Publish : OK"));
	}
	else 
	{
		Serial.println( F("Publish : FAIL!!!"));
		Serial.print( F("Wifi State is : "));
		Serial.println(wifiManager.state());
		Serial.println( F("Deconnecting WiFi!"));
		wifiManager.disconnect();
		delay(500);
		Serial.println( F("Wifi State after deconnecting is : "));
		Serial.print(wifiManager.state());
		Serial.println( F("Re-connecting WiFi!"));
		wifiManager.begin();
		Serial.println( F("Wifi State after re-connecting is : "));
		Serial.print(wifiManager.state());
		begin();
	}
	return(pub);

}


/**
*	CoolMQTT::mqttLoop():
*	This method is provided to allow the
*	client to process the data
*
*	\return true if successful,false
*	otherwise
*/	
bool CoolMQTT::mqttLoop()
{

	unsigned long lastTime=millis();

#if DEBUG == 1

	Serial.println( F("Entering CoolMQTT.mqttLoop()") );
	Serial.println();

#endif	

	while( ( millis() - lastTime ) < 1000)
	{
		this->client.loop();
		yield();
	}

#if DEBUG == 1 
	
	Serial.print( F("loop result : ") );
	Serial.println( this->client.loop() );
	Serial.println();

#endif

	return( this->client.loop() );
}

/**
*	CoolMQTT::callback(in topic, incoming message , message length):
*	This method is provided to handle incoming messages from the
*	subscribed inTopic.
*	
*	Arguments are automatically assigned in client.setCallback()
*/
void CoolMQTT::callback(char* topic, byte* payload, unsigned int length) 
{

#if DEBUG == 1

	Serial.println( F("Entering CoolMQTT.callback() ") );
	Serial.println();

#endif 

	if(this->newMsg==false)
	{
		char temp[length+1];

	#if DEBUG == 1

		Serial.println( F("received temp msg : ") );
		
	#endif
		
		for (unsigned int i = 0; i < length; i++) 
		{
			temp[i]=(char)payload[i];
		
		#if DEBUG == 1 

			Serial.print( (char)payload[i] );
		
		#endif

		}
	
	#if DEBUG == 1 

		Serial.println();
		Serial.println( F("storing new message : ") );

		Serial.print(F("length : "));
		Serial.println(length);
		
		Serial.print(F("size : "));
		Serial.print(sizeof(payload));
		Serial.println();
	
	#endif

		this->newMsg=true;

		temp[length+1]='\0';

		this->msg=String(temp);
		this->msg.remove(length,1);
	
	#if DEBUG == 1 

		Serial.println( F("stored message : ") );
		Serial.println(this->msg);
	
	#endif

	}
	else
	{
	
	#if DEBUG == 1

		Serial.println( F("did not read last message") );
	
	#endif 
		
	}

}

/**
*	CoolMQTT::read():
*	This method is provided to return the last
*	read message.
*/
String CoolMQTT::read()
{	

#if DEBUG == 1 

	Serial.println( F("Entering CoolMQTT.read()") );
	Serial.println();

#endif 

	if(this->newMsg==true)
	{
		
		this->newMsg=false;

#if DEBUG == 1 
		Serial.println( F("received new message") );
		Serial.println( F("message : ") );
		Serial.println(this->msg);
		Serial.println();

#endif

		return(this->msg);
		
	}
	return("");

}

/**
*	CoolMQTT::config():
*	This method is provided to configure
*	the mqttClient :	-server
*				-inTopic
*				-outTopic
*				-client Id
*				-buffer size	
*
*	\return true if successful,false otherwise
*/
bool CoolMQTT::config()
{

#if DEBUG == 1 

	Serial.println( F("Entering CoolMQTT.config()") );
	Serial.println();

#endif

	//read config file
	//update data
	File configFile = SPIFFS.open("/mqttConfig.json", "r");

	if (!configFile) 
	{
	
	#if DEBUG == 1 

		Serial.println( F("failed to read /mqttConfig.json") );
		Serial.println();

	#endif

		return(false);
	}
	else
	{
		size_t size = configFile.size();
		// Allocate a buffer to store contents of the file.
		std::unique_ptr<char[]> buf(new char[size]);

		configFile.readBytes(buf.get(), size);
		DynamicJsonBuffer jsonBuffer;
		JsonObject& json = jsonBuffer.parseObject(buf.get());
		if (!json.success()) 
		{
		
		#if DEBUG == 1 

			Serial.println( F("failed to parse json ") );
			Serial.println();
		
		#endif
			
			return(false);
		} 
		else
		{
		
		#if DEBUG == 1 
		
			Serial.println( F("configuration json is ") );
			json.printTo(Serial);
			Serial.println();

			Serial.print(F("jsonBuffer size: "));
			Serial.println(jsonBuffer.size());
			Serial.println();


		#endif

			if(json["mqttServer"].success() )
			{			
				const char* tempmqttServer = json["mqttServer"]; 
				for(int i =0;i< 50 ;i++)
				{
					mqttServer[i]=tempmqttServer[i];
				}
			}
			else
			{
				for(int i =0;i< 50 ;i++)
				{
					this->mqttServer[i]=this->mqttServer[i];
				}

			}
			json["mqttServer"]=this->mqttServer;

			
			if(json["inTopic"].success() )
			{
				const char* tempInTopic = json["inTopic"]; 
				for(int i =0;i< 50;i++)
				{
					inTopic[i]=tempInTopic[i];
				}
			}
			else
			{
				String tempMAC = WiFi.macAddress();
				tempMAC.replace(":","");
				snprintf(inTopic, 50, "$aws/things/%s/shadow/update/delta", tempMAC.c_str());	
			
			#if DEBUG == 1				
				
				Serial.print( F("Set Incomming MQTT Channel to : ") );
				Serial.println(inTopic);
			
			#endif	

			}
			json["inTopic"]=this->inTopic;
			
			
			if(json["outTopic"].success() )
			{
				const char* tempOutTopic = json["outTopic"]; 
				for(int i =0;i<50;i++)
				{
					outTopic[i]=tempOutTopic[i];
				}
			}
			else
			{
				String tempMAC = WiFi.macAddress();
				tempMAC.replace(":","");
				snprintf(outTopic, 50, "$aws/things/%s/shadow/update", tempMAC.c_str());
			
			#if DEBUG == 1 

				Serial.print( F("Set Outgoing MQTT Channel to : ") );
				Serial.println(outTopic);
			
			#endif

			}
			json["outTopic"]=this->outTopic;
		
			
			if(json["user"].success() )
			{				
				const char* tempUser = json["user"]; 
				for(int i =0;i<50;i++)
				{
					user[i]=tempUser[i];
				}
			}
			else
			{
				for(int i=0;i<50;i++)
				{
					this->user[i]=this->user[i];
				}				
			}
			json["user"]=this->user;
			
			if(json["bufferSize"].success() )
			{
				int tempBufferSize = json["bufferSize"]; 
				bufferSize=tempBufferSize;
			}
			else
			{
				this->bufferSize=this->bufferSize;
			}
			json["bufferSize"]=this->bufferSize;

			configFile.close();
			configFile = SPIFFS.open("/mqttConfig.json", "w");
			if(!configFile)
			{
			
			#if DEBUG == 1 

				Serial.println( F("failed to write to /mqttConfig.json") );
			
			#endif

				return(false);				
			}
			
			json.printTo(configFile);
			configFile.close();

		#if DEBUG == 1 

			Serial.println( F("saved configuration is :") );
			json.printTo(Serial);
			Serial.println();
		
		#endif

			return(true); 
		}
	}	
	

}

/**
*	CoolMQTT::config(server,in topic, out topic , user Id, buffer size):
*	This method is provided to manually configure the mqtt client	
*
*/
void CoolMQTT::config(const char mqttServer[],const char inTopic[],const char outTopic[],const char user[],int bufferSize)
{

#if DEBUG == 1

	Serial.println( F("Entering CoolMQTT.config() , no SPIFFS variant") );
	Serial.println();

#endif

	for(int i =0;i< 50 ;i++)
	{
		this->mqttServer[i]=mqttServer[i];
		this->inTopic[i]=inTopic[i];
		this->outTopic[i]=outTopic[i];
		this->user[i]=user[i];
	}
	this->bufferSize=bufferSize;
	

}

/**
*	CoolMQTT::printConf():
*	This method is provided to print the
*	configuration to the Serial Monitor
*/
void CoolMQTT::printConf()
{

#if DEBUG == 1 

	Serial.println( F("Entering CoolMQTT.printConf()") );
	Serial.println();	

#endif
	
	Serial.println("MQTT configuration ");

	Serial.print("mqttServer : ");
	Serial.println(this->mqttServer);

	Serial.print("inTopic : ");
	Serial.println(this->inTopic);

	Serial.print("outTopic : ");
	Serial.println(this->outTopic);

	Serial.print("user : ");
	Serial.println(this->user);

	Serial.print("bufferSize : ");
	Serial.println(this->bufferSize);

	Serial.println();


}

/**
*	CoolMQTT::getUser():
*	This method is provided to get the user name
*/
String CoolMQTT::getUser()
{

#if DEBUG == 1 
	Serial.println( F("Entering CoolMQTT.getUser()") );
	Serial.println();
	
	Serial.print( F("user : ") );
	Serial.println(this->user);

#endif

	return String(this->user);
}
